bokeh/document.py

```diff
@@ -25,6 +26,38 @@
 
 DEFAULT_TITLE = "Bokeh Application"
 
+class DocumentProxy(object):
+    ''' Wrap a Document object so that only methods that can safely be used
+    from unlocked callbacks or threads are exposed. Attempts to otherwise
+    access or change the Document results in an exception.
+
+    '''
+
+    def __init__(self, doc):
+        self._doc = doc
+
+    def __getattr__(self, attr):
+        raise RuntimeError("Only add_next_tick_callback may be used safely from DocumentProxy")
```

Comment:
this seems like it could be clearer, like "Only add_next_tick_callback may be used safely without taking the document lock; to make other changes to the document, add a next tick callback and make your changes from that callback." - DocumentProxy is sort of an implementation detail it seems like.

Something I just realized is that we pretty much encourage saving models in variables, and touching the models directly, and people will often do that, so this warning isn't going to catch everything by any means. But, better than nothing.